File: LeetCode/Python Solutions/Sort Items by Groups Respecting Dependencies/sort.py
from collections import defaultdict
from typing import List, Dict

class Solution:
    def sortItems(self, n: int, m: int, group: List[int], beforeItems: List[List[int]]) -> List[int]:
        # Assign new groups for items without a group.
        # Start new group ids from m as [0 to m-1] already used.
        for i in range(n):
            if group[i] == -1:
                group[i] = m
                m += 1  # increase the group count
        
        # Create an adjacency list (dependency graph) for items
        item_graph = defaultdict(list)
        # In-degree count for all items. Used for topological sort.
        item_indegree = [0] * n
        # Create an adjacency list (dependency graph) for groups
        group_graph = defaultdict(list)
        # In-degree count for groups. Used for topological sort.
        group_indegree = [0] * m
        
        # Construct the graphs based on beforeItems list
        for i in range(n):
            for pre_item in beforeItems[i]:
                item_graph[pre_item].append(i)
                item_indegree[i] += 1
                
                # If the two items 'i' and 'pre_item' are from different groups, 
                # add an edge between these groups.
                if group[i] != group[pre_item]:
                    group_graph[group[pre_item]].append(group[i])
                    group_indegree[group[i]] += 1
        
        # Topological sort on items and groups
        item_order = self.topological_sort(item_graph, item_indegree)
        group_order = self.topological_sort(group_graph, group_indegree)
        
        # If we can't get a valid order, return []
        if not item_order or not group_order:
            return []
        
        # Arrange items within each group in the sorted order
        group_to_items = defaultdict(list)
        for item in item_order:
            group_to_items[group[item]].append(item)
        
        # Combine items from all groups in the order of sorted groups.
        sorted_items = []
        for group_id in group_order:
            sorted_items.extend(group_to_items[group_id])
        
        return sorted_items
    
    def topological_sort(self, graph: Dict[int, List[int]], indegree: List[int]) -> List[int]:
        # Initialize a result list and a stack to keep nodes with 0 in-degree
        sorted_order = []
        zero_indegree = [i for i, deg in enumerate(indegree) if deg == 0]
        
        # While we have nodes with 0 in-degree, process them
        while zero_indegree:
            node = zero_indegree.pop()  # get a node with 0 in-degree
            sorted_order.append(node)   # add to result list
            # Decrease in-degree for all neighbors
            for neighbor in graph[node]:
                indegree[neighbor] -= 1
                # If neighbor becomes 0 in-degree, add to stack
                if indegree[neighbor] == 0:
                    zero_indegree.append(neighbor)
        
        # Return sorted order if we have processed all nodes; otherwise, return []
        return sorted_order if len(sorted_order) == len(graph) else []

File: LeetCode/Java Solutions/Sort Items by Groups Respecting Dependencies/sort.java
import java.util.*;

public class Solution {
    
    public int[] sortItems(int n, int m, int[] group, List<List<Integer>> beforeItems) {
        
        // Assign new groups for items without a group.
        for (int i = 0; i < n; i++) {
            if (group[i] == -1) {
                group[i] = m;
                m++; // Increase the group count
            }
        }
        
        // Create an adjacency list (dependency graph) for items
        List<Integer>[] itemGraph = new ArrayList[n];
        for (int i = 0; i < n; i++) itemGraph[i] = new ArrayList<>();
        
        // In-degree count for all items. Used for topological sort.
        int[] itemInDegree = new int[n];
        
        // Create an adjacency list (dependency graph) for groups
        List<Integer>[] groupGraph = new ArrayList[m];
        for (int i = 0; i < m; i++) groupGraph[i] = new ArrayList<>();
        
        // In-degree count for groups. Used for topological sort.
        int[] groupInDegree = new int[m];
        
        // Construct the graphs based on beforeItems list
        for (int i = 0; i < n; i++) {
            for (int preItem : beforeItems.get(i)) {
                itemGraph[preItem].add(i);
                itemInDegree[i]++;
                
                if (group[i] != group[preItem]) {
                    groupGraph[group[preItem]].add(group[i]);
                    groupInDegree[group[i]]++;
                }
            }
        }
        
        // Topological sort on items and groups
        List<Integer> itemOrder = topologicalSort(itemGraph, itemInDegree);
        List<Integer> groupOrder = topologicalSort(groupGraph, groupInDegree);
        
        // If we can't get a valid order, return []
        if (itemOrder.size() != n || groupOrder.size() != m) return new int[0];
        
        // Arrange items within each group in the sorted order
        Map<Integer, List<Integer>> groupToItems = new HashMap<>();
        for (int item : itemOrder) {
            groupToItems.putIfAbsent(group[item], new ArrayList<>());
            groupToItems.get(group[item]).add(item);
        }
        
        // Combine items from all groups in the order of sorted groups.
        int[] sortedItems = new int[n];
        int idx = 0;
        for (int groupId : groupOrder) {
            if (groupToItems.containsKey(groupId)) {  // Check if group has items
                for (int item : groupToItems.get(groupId)) {
                    sortedItems[idx++] = item;
                }
            }
        }
        
        return sortedItems;
    }
    
    private List<Integer> topologicalSort(List<Integer>[] graph, int[] indegree) {
        List<Integer> sortedOrder = new ArrayList<>();
        Queue<Integer> zeroInDegree = new LinkedList<>();
        
        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0) zeroInDegree.add(i);
        }
        
        while (!zeroInDegree.isEmpty()) {
            int node = zeroInDegree.poll();
            sortedOrder.add(node);
            for (int neighbor : graph[node]) {
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) zeroInDegree.add(neighbor);
            }
        }
        
        return sortedOrder;
    }
}
